designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,420 @@
+    let promise
+    try {
+      const report = this.cliEngine.executeOnFiles(patterns)
+      promise = Promise.resolve(report.results)
+    } catch (error) {
+      promise = Promise.reject(error)
+    }
+
+    return {
+      then(onFulfilled, onRejected) {
+        return promise.then(onFulfilled, onRejected)
+      },
+      async *[Symbol.asyncIterator]() {
+        yield* await promise
+      },
+    }
+  }
+}
+```
+
+</details>
+
+#### § The `executeOnText()` method
+
+This method returns the same type of an object as the `executeOnFiles()` method.
+
+Because the returned object of `CLIEngine#executeOnText()` method is the same type as the `CLIEngine#executeOnFiles()` method. The `ESLint` class inherits that mannar.
+
+<details>
+<summary>Example: Show the result.</summary>
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+
+const [result] = await eslint.executeOnText(text, filePath)
+print(result)
+```
+
+</details>
+
+<details>
+<summary>Example: Using along with the `executeOnFiles()` method.</summary>
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+
+const report = useStdin
+  ? eslint.executeOnText(text, filePath)
+  : eslint.executeOnFiles(patterns)
+
+for await (const result of report) {
+  print(result)
+}
+```
+
+</details>
+
+#### § The `getFormatter()` method
+
+This method returns a `Promise<Formatter>`. The `Formatter` type is a function `(results: AsyncIterable<LintResult>) => AsyncIterableIterator<string>`. It receives lint results then outputs the formatted text.
+
+This means the `getFormatter()` method wraps the current formatter to adapt the interface.
+
+<details>
+<summary>A rough sketch of the `getFormatter()` method.</summary>
+
+```js
+class ESLint {
+  async getFormatter(name) {
+    const format = this.cliEngine.getFormatter(name)
+
+    // Return the wrapper.
+    return async function* formatter(resultIterator) {
+      // Collect and sort the results.
+      const results = []
+      for await (const result of resultIterator) {
+        results.push(result)
+      }
+      results.sort(ESLint.compareResultsByFilePath)
+
+      // Make `rulesMeta`.
+      const rules = this.cliEngine.getRules()
+      const rulesMeta = getRulesMeta(rules)
+
+      // Format the results with the formatter of the current spec.
+      yield format(results, { rulesMeta })
+    }
+  }
+}
+```
+
+</details>
+
+<details>
+<summary>Example: Use the formatter.</summary>
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+const formatter = eslint.getFormatter("stylish")
+
+// Verify files
+const results = eslint.executeOnFiles(patterns)
+// Format and write the results
+for await (const textPiece of formatter(results)) {
+    process.stdout.write(textPiece)
+}
+```
+
+</details>
+
+Once the `getFormatter()` method got this change, we can update the specification of custom formatters without breakage in the future to support streaming.
```

Comment:
See my previous comment. Should probably be more fleshed out the intention of why this is an async generator and how that would work right here and now and how that would work eventually in the future.

It should probably also be added some reasoning on why the formatters are okay with being pure async generators while `executeOnFiles` has been made a cross-product of async generators and promises. Should the two be consistent in that regard?

Reply:
> Should the two be consistent in that regard?

Because formatters are used to output results to stdout or a file, and both are streams, so I didn't think that it should be Thenable.